hooks/useTransactionSender.tsx

```diff
@@ -1,25 +1,320 @@
+  transaction: TransactionWithMetadata<T>;
+  result?: SignatureResult;
+  status?: TransactionStatus;
+};
+
+// type guard function function for TransactionWithMetadata since instanceof won't work
+const isTransactionWithMetadata = <T extends Transaction | VersionedTransaction>(
+  obj: any,
+): obj is TransactionWithMetadata<T> => {
+  return typeof obj === 'object' && obj !== null && 'tx' in obj;
+};
+
+export const useTransactionSender = <T extends Transaction | VersionedTransaction>() => {
   const { connection } = useConnection();
   const wallet = useWallet();
   const { priorityFee } = usePriorityFee();
+  const { generateExplorerLink } = useExplorerConfiguration();
+
+  const [idToTransactionInfos, setIdToTransactionInfos] = useState<
+    Record<string, Array<TransactionInfo<T>>>
+  >({});
+  const [successfulSignatureCount, setSuccessfulSignatureCount] = useState<Record<string, number>>(
+    {},
+  );
+  const [idsToClear, setIdsToClear] = useState<Array<string>>([]);
+
+  useEffect(() => {
+    asyncClearStateForNotificationId();
+
+    return () => {
+      setIdsToClear([]);
+    };
+  }, [idsToClear]);
+
+  /**
+   * render a notification update when a transaction result is received
+   */
+  useEffect(() => {
+    const notificationIds = Object.keys(idToTransactionInfos);
+    if (notificationIds.length === 0) return;
+
+    notificationIds.forEach((id) => {
+      const transactionInfos = idToTransactionInfos[id];
+
+      notifications.update({
+        id,
+        title: <Text fw="bold">{generateTitle(transactionInfos, id)}</Text>,
+        message: generateNotficationBody(transactionInfos),
+      });
+    });
+  }, [idToTransactionInfos]);
+
+  const updateSingleTransactionInfo = (
+    signature: string,
+    id: string,
+    updateStateFieldsCallback: () => Partial<TransactionInfo<T>>,
+  ) => {
+    setIdToTransactionInfos((state) => {
+      if (!state[id]) return state;
+      const indexToUpdate = state[id].findIndex((x) => x.signature === signature);
+
+      // no transaction info for id, ignore
+      if (indexToUpdate === -1) return state;
+      return {
+        ...state,
+        [id]: [
+          ...(indexToUpdate === 0 ? [] : state[id].slice(0, indexToUpdate)),
+          {
+            ...state[id][indexToUpdate],
+            ...updateStateFieldsCallback(),
+          },
+          // start index is inclusive, add 1 to skip the updated element
+          ...(indexToUpdate + 1 >= state[id].length ? [] : state[id].slice(indexToUpdate + 1)),
+        ],
+      };
+    });
+  };
+
+  /**
+   * all state that we process in this function wasn't yet updated. i failed
+   * to figure out exactly what was going on, but my problems were solved by pushing new state
+   * and allowing async state updates to trigger state cleanup logic
+   */
+  const onNotificationClose = (id: string) => setIdsToClear((state) => [...state, id]);
+
+  const asyncClearStateForNotificationId = () => {
+    idsToClear.forEach((id) => {
```

Comment:
How do we set this? Ideally I would say let's pass this array into the function so it's a more pure function.

Also on this function name, I noticed it starts with `async` but I'm not seeing anything async unless my eyes deceive me...

Reply:
New IDs get added to the array when a user clicks on the close notification button, and we trigger this function in a useEffect when the array changes. Ideally this would happen in the `onClick` callback, but I had trouble with the latest state being up-to-date, and invoking this async after state updates was the only way I could get it to work.

Ah, my bad - poor naming convention on my part. I named it async due to the fact that it gets invoked after asynchronous state updates in React. But, it is confusing. I can rename.

Reply:
Ahh okay I see now what you were getting at. Well yeah maybe a name like `clearStateForNotificationIds` since it does multiple instead of just one? How does that sound? Or whatever you think too.

Reply:
Yeah, after your comment about using a pure function, I thought maybe it was clearer by just putting it in useEffect. Although, I am happy to move it back into its own function if you disagree!

Reply:
@jshiohaha naw I like that approach. Looks good to me!